twoliter/embedded/imghelper

```diff
@@ -0,0 +1,398 @@
+      -export \
+      -passout pass: \
+      -inkey "${SBKEYS}/shim-sign.key" \
+      -in "${SBKEYS}/shim-sign.crt" \
+      -certfile "${SBKEYS}/db.crt" \
+      -out "${SBKEYS}/shim-sign.p12"
+
+    openssl pkcs12 \
+      -export \
+      -passout pass: \
+      -inkey "${SBKEYS}/code-sign.key" \
+      -in "${SBKEYS}/code-sign.crt" \
+      -certfile "${SBKEYS}/vendor.crt" \
+      -out "${SBKEYS}/code-sign.p12"
+
+    # Import certificates and private key archive.
+    pedb="/etc/pki/pesign"
+
+    certutil -d "${pedb}" -A -n db -i "${SBKEYS}/db.crt" -t ",,C"
+    certutil -d "${pedb}" -A -n shim-sign-key -i "${SBKEYS}/shim-sign.crt" -t ",,P"
+    pk12util -d "${pedb}" -i "${SBKEYS}/shim-sign.p12" -W ""
+
+    certutil -d "${pedb}" -A -n vendor -i "${SBKEYS}/vendor.crt" -t ",,C"
+    certutil -d "${pedb}" -A -n code-sign-key -i "${SBKEYS}/code-sign.crt" -t ",,P"
+    pk12util -d "${pedb}" -i "${SBKEYS}/code-sign.p12" -W ""
+
+    certutil -d "${pedb}" -L
+    sb_key_source="local"
+    shim_sign_key=(-c shim-sign-key)
+    code_sign_key=(-c code-sign-key)
+  fi
+
+  # Convert certificates from PEM format (ASCII) to DER (binary). This could be
+  # done when the certificates are created, but the resulting binary files are
+  # not as nice to store in source control.
+  for cert in PK KEK db vendor; do
+    openssl x509 \
+      -inform PEM -in "${SBKEYS}/${cert}.crt" \
+      -outform DER -out "${SBKEYS}/${cert}.cer"
+  done
+
+  # For signing the grub config, we need to embed the GPG public key in binary
+  # form, which is similarly awkward to store in source control.
+  gpg --batch --no-tty --import "${SBKEYS}/config-sign.key"
+  if [[ "${sb_key_source}" == "aws" ]]; then
+    gpg --batch --no-tty --card-status
+  fi
+  gpg --batch --no-tty --export >"${SBKEYS}/config-sign.pubkey"
+  gpg --batch --no-tty --list-keys
+}
+
+replace_vendor_certs() {
+  local -n shim_sign_key code_sign_key
+  local uefi_secure_boot
+  shim_sign_key="${1:?}"
+  code_sign_key="${2:?}"
+  uefi_secure_boot="${3:?}"
+  shims=(boot*.efi)
+  shim="${shims[0]}"
+  grubs=(grub*.efi)
+  grub="${grubs[0]}"
+  mokms=(mm*.efi)
+  mokm="${mokms[0]}"
+  if [[ "${uefi_secure_boot}" == "yes" ]]; then
+    # Convert the vendor certificate to the expected format.
+    cert_table "${SBKEYS}/vendor.cer" "${SBKEYS}/vendor.obj"
+
+    # Replace the embedded vendor certificate, then sign shim with the db key.
+    pesign -i "${shim}" -l
+    objdumpcopy "${shim}" \
+      --update-section ".vendor_cert=${SBKEYS}/vendor.obj"
+    pesign -i "${shim}" -o "${shim}.signed" -s "${shim_sign_key[@]}"
+    mv "${shim}.signed" "${shim}"
+    pesign -i "${shim}" -l
+    pesigcheck -i "${shim}" -n 0 -c "${SBKEYS}/db.cer"
+
+    # Remove the original key if necessary, then sign the MOK manager as well.
+    if [[ "$(pesign -i "${mokm}" -l)" == "No signatures found." ]]; then
+      mv "${mokm}" "${mokm}.unsigned"
+    else
+      pesign -i "${mokm}" -o "${mokm}.unsigned" -u 0 -r
+    fi
```

Comment:
Why is this necessary for mokm and grub but not for shim? It would be cleaner to treat all the same way.

Reply:
After running `objdumpcopy` the signature disappears, but I can add the remove before the `objdumpcopy` to be safe and consistent.